datasource/src/main/scala/quasar/plugin/avalanche/datasource/AvalancheLoader.scala

```diff
@@ -0,0 +1,155 @@
+
+import java.net.InetAddress
+import java.nio.ByteBuffer
+import java.sql.ResultSet
+import java.time._
+import java.util.UUID
+
+import cats.effect.Resource
+import cats.implicits._
+
+import doobie._
+import doobie.enum.JdbcType
+import doobie.implicits._
+
+import fs2.Stream
+
+import quasar.ScalarStages
+import quasar.common.data.{QDataRValue, RValue}
+import quasar.connector.QueryResult
+import quasar.connector.datasource.BatchLoader
+import quasar.plugin.jdbc._
+import quasar.plugin.jdbc.datasource._
+
+private[datasource] object AvalancheLoader {
+  type I = AvalancheHygiene.HygienicIdent
+  type Args = (I, Option[I], ColumnSelection[I], ScalarStages)
+
+  def apply(logHandler: LogHandler, resultChunkSize: Int)
+      : BatchLoader[Resource[ConnectionIO, ?], Args, QueryResult[ConnectionIO]] =
+    BatchLoader.Full[Resource[ConnectionIO, ?], Args, QueryResult[ConnectionIO]] {
+      case (table, schema, columns, stages) =>
+        val dbObject0 =
+          schema.fold(table.fr0)(_.fr0 ++ Fragment.const0(".") ++ table.fr0)
+
+        val projections = Some(columns) collect {
+          case ColumnSelection.Explicit(idents) =>
+            idents.map(_.fr0).intercalate(fr",")
+
+          case ColumnSelection.All => fr0"*"
+        }
+
+        val rvalues = projections match {
+          case Some(prjs) =>
+            val sql =
+              (fr"SELECT" ++ prjs ++ fr" FROM" ++ dbObject0).query[Unit].sql
+
+            val ps =
+              FC.prepareStatement(
+                sql,
+                ResultSet.TYPE_FORWARD_ONLY,
+                ResultSet.CONCUR_READ_ONLY)
+
+            loggedRValueQuery(sql, ps, resultChunkSize, logHandler)(isSupported, unsafeRValue)
+
+          case None =>
+            (Stream.empty: Stream[ConnectionIO, RValue]).pure[Resource[ConnectionIO, ?]]
+        }
+
+        rvalues.map(QueryResult.parsed(QDataRValue, _, stages))
+    }
+
+  def isSupported(sqlType: SqlType, avalancheType: VendorType): Boolean =
+    SupportedSqlTypes(sqlType) || Types.AvalancheColumnTypes.contains(avalancheType)
+
+  def unsafeRValue(rs: ResultSet, col: ColumnNum, sqlType: SqlType, vendorType: VendorType): RValue = {
+    import java.sql.Types._
+
+    def unlessNull[A](a: A)(f: A => RValue): RValue =
+      if (a == null) null else f(a)
+
+    (sqlType: @switch) match {
+      case CHAR | NCHAR | NVARCHAR =>
+        unlessNull(rs.getString(col))(RValue.rString(_))
+
+      case VARCHAR =>
+        unlessNull(rs.getString(col)) { s =>
+          if (vendorType == Types.IntervalYearToMonth)
+            RValue.rInterval(IntervalParser.unsafeParseYearToMonth(s))
+          else if (vendorType == Types.IntervalDayToSecond)
+            RValue.rInterval(IntervalParser.unsafeParseDayToSecond(s))
+          else
+            RValue.rString(s)
+        }
+
+      case TINYINT | SMALLINT | INTEGER | BIGINT =>
+        unlessNull(rs.getLong(col))(RValue.rLong(_))
+
+      case DOUBLE | FLOAT | REAL =>
```

Comment:
What's `REAL`? it should be a JVM `Double` and not `BigDecimal`?

Reply:
It is a [4-byte floating point number](https://docs.actian.com/avalanche/index.html#page/SQLLanguage%2F2._SQL_Data_Types.htm%23ww414919), so `Double` should be ok.